### PR TITLE
refactor(utils): Update gear-replay-cli tool

### PR DESCRIPTION
**Pull request by ekovalev**
Aligns our custom `gear-replay-cli` tool with changes in underlying Substrate code. Also, following the original `try-runtime-cli` having been placed in the standalone repo under Parity Tech, synced our tool with its original predecessor.

More specifically,

- added support for state snapshots (which is useful for large states like our testnet, for instance, which is over 1.6GB at this point);
- fixed an error with forcing `gear::run()` extrinsic in a block: the filter based on comparing encoded extrinsics was unstable.

**Note:** the command format has naturally changed a bit - refer to the examples in README.md for the up to date syntax.

### CODE REVIEW

File: utils/gear-replay-cli/src/cmd/create_snapshot.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                remote_spec.spec_name.to_lowercase(),
+                remote_spec.spec_version,
+                block_hash,
+            );
+            log::info!(target: LOG_TARGET, "snapshot path not provided (-s), using '{}'", path_str);
+            path_str
+        }
+    };
+
+    // // Create executor, suitable for usage in conjunction with the preferred execution strategy.
```
Comment (NikVolf) [inline utils/gear-replay-cli/src/cmd/create_snapshot.rs:77]: remove commented?

File: utils/gear-replay-cli/src/lib.rs [hunk 2]
```diff
@@ -16,103 +16,262 @@
...
+        .with_onchain_heap_alloc_strategy(heap_pages)
+        .with_offchain_heap_alloc_strategy(heap_pages)
+        .with_max_runtime_instances(max_runtime_instances)
+        .with_runtime_cache_size(runtime_cache_size)
+        .build()
 }
 
-pub async fn run() -> sc_cli::Result<()> {
-    gear_runtime_interface::sandbox_init();
+// pub(crate) async fn build_externalities<Block: BlockT + DeserializeOwned>(
```
Comment (ukint-vs) [inline utils/gear-replay-cli/src/lib.rs:142]: remove?

File: utils/gear-replay-cli/README.md [hunk 3]
```diff
@@ -4,174 +4,124 @@
...
+    -u, --uri <URI>                    The RPC url [default: wss://archive-rpc.vara.network:443]
+    -b, --block <BLOCK>                The block hash or number we want to replay. If omitted, the latest finalized block is used. The blockchain state at previous block with respect to this parameter will be scraped
+    -p, --pallet <PALLET>...           Pallet(s) to scrape. Comma-separated multiple items are also accepted. If empty, entire chain state will be scraped
+        --prefix <HASHED_PREFIXES>...  Storage entry key prefixes to scrape and inject into the test externalities. Pass as 0x prefixed hex strings. By default, all keys are scraped and included
+        --child-tree                   Fetch the child-keys as well
+    -h, --help                         Print help (see more with '--help')
+```
+Usage example:
+```bash
+export BLOCK_NUM=2000000
```
Comment (gshep) [inline utils/gear-replay-cli/README.md:77]: If I'm not mistaken there is no mention about the env var in the help of create-snapshot subcommand. Would be nice to write a couple of words about that.

And am I right that the `-b`/`--block <BLOCK>` arguments have the priority over the environment variable?

File: utils/gear-replay-cli/src/cmd/replay_block.rs [hunk 4]
```diff
@@ -181,8 +179,8 @@ where
         state_machine_call::<Block, _>(&ext, &executor, method, &payload, full_extensions())?;
     log::info!(
         target: LOG_TARGET,
-        "Core_execute_block for block {} completed",
-        current_number
+        "Core_execute_block for block {:?} completed",
```
Comment (gshep) [inline utils/gear-replay-cli/src/cmd/replay_block.rs:182]:
```suggestion
        "Core_execute_block for block {current_hash:?} completed",
```

File: utils/gear-replay-cli/src/lib.rs [hunk 5]
```diff
@@ -16,103 +16,262 @@
...
     Number(NumberFor<B>),
 }
 
-/// Commands of `gear-replay` CLI
-#[derive(Debug, Subcommand)]
-pub enum Command {
-    ReplayBlock(replay_block::ReplayBlockCmd<Block>),
-    GearRun(gear_run::GearRunCmd<Block>),
+impl<B: BlockT> core::fmt::Display for BlockHashOrNumber<B> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
```
Comment (gshep) [inline utils/gear-replay-cli/src/lib.rs:63]:
```suggestion
impl<B: BlockT> fmt::Display for BlockHashOrNumber<B> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

and use `std::fmt::{self, Debug};`

File: utils/gear-replay-cli/src/lib.rs [hunk 6]
```diff
@@ -16,103 +16,262 @@
...
-struct ReplayCli {
-    #[clap(flatten)]
-    pub shared: SharedParams,
+#[cfg(not(feature = "always-wasm"))]
+pub(crate) fn build_executor<D: NativeExecutionDispatch>() -> NativeElseWasmExecutor<D> {
+    let heap_alloc_strategy = HeapAllocStrategy::Dynamic {
+        maximum_pages: Some(2048),
+    };
+    let max_runtime_instances = 8;
+    let runtime_cache_size = 2;
```
Comment (gshep) [inline utils/gear-replay-cli/src/lib.rs:108]: Maybe it is worth to get the values from arguments? Insubstantial, `todo` and an issue for future is more than enuogh

File: utils/gear-replay-cli/src/parse.rs [hunk 7]
```diff
@@ -21,6 +21,23 @@
...
 use hex::FromHexError;
 
+pub(crate) fn hash(block_hash: &str) -> Result<String, String> {
+    let (block_hash, offset) = if let Some(block_hash) = block_hash.strip_prefix("0x") {
+        (block_hash, 2)
+    } else {
+        (block_hash, 0)
+    };
+
+    if let Some(pos) = block_hash.chars().position(|c| !c.is_ascii_hexdigit()) {
```
Comment (gshep) [inline utils/gear-replay-cli/src/parse.rs:31]: perhaps `match` is more appropriate for both cases in this func

**Review by gshep — APPROVED** [review]
Good job! :1st_place_medal:

File: utils/gear-replay-cli/README.md (hunk 3 above)
Comment (ekovalev) [reply utils/gear-replay-cli/README.md:77]: Maybe the doc is a bit misleading: the tool doesn't check any env vars - only command line arguments. It's just in the doc itself I use these BLOCK_NUM and HASH across a number of example commands so I figured they should be defined somewhere.
Apparently this causes ambiguity. Let me remove the env vars from the doc completely.

File: utils/gear-replay-cli/src/lib.rs (hunk 6 above)
Comment (ekovalev) [reply utils/gear-replay-cli/src/lib.rs:108]: That's what the original `try-runtime-cli` does now. I thought it was insignificant for us at this point, but can just copy and paste a few extra lines to take these values as CLI arguments.

Comment (gshep) [reply utils/gear-replay-cli/src/lib.rs:108]: ok

File: utils/gear-replay-cli/src/parse.rs (hunk 7 above)
Comment (ekovalev) [reply utils/gear-replay-cli/src/parse.rs:31]: A matter of preference, I guess. Both versions seem to produce identical code when compiled.

**Pull request merged**